code/modules/station_goals/bluespace_tap.dm

```diff
@@ -0,0 +1,429 @@
+		input_level--
+		return	// and no mining gets done
+	if(actual_power_usage)
+		add_load(actual_power_usage)
+		var/points_to_add = (input_level + emagged) * base_points
+		points += points_to_add	//point generation, emagging gets you 'free' points at the cost of higher anomaly chance
+		total_points += points_to_add
+	// actual input level changes slowly
+	if(input_level < desired_level && (surplus() >= get_power_use(input_level + 1)))
+		input_level++
+	else if(input_level > desired_level)
+		input_level--
+	if(prob(input_level - safe_levels + (emagged * 5)))	//at dangerous levels, start doing freaky shit. prob with values less than 0 treat it as 0
+		GLOB.event_announcement.Announce("Unexpected power spike during Bluespace Harvester Operation. Extra-dimensional intruder alert. Expected location: [get_area(src).name]. [emagged ? "DANGER: Emergency shutdown failed! Please proceed with manual shutdown." : "Emergency shutdown initiated."]", "Bluespace Harvester Malfunction")
+		if(!emagged)
+			input_level = 0	//emergency shutdown unless we're sabotaged
+			desired_level = 0
+		for(var/i in 1 to rand(1, 3))
+			var/turf/location = locate(x + rand(-5, 5), y + rand(-5, 5), z)
+			new /obj/structure/spawner/nether/bluespace_tap(location)
+
+
+
+/obj/machinery/power/bluespace_tap/tgui_data(mob/user)
+	var/data[0]
+
+	data["desiredLevel"] = desired_level
+	data["inputLevel"] = input_level
+	data["points"] = points
+	data["totalPoints"] = total_points
+	data["powerUse"] = actual_power_usage
+	data["availablePower"] = surplus()
+	data["maxLevel"] = max_level
+	data["emagged"] = emagged
+	data["safeLevels"] = safe_levels
+	data["nextLevelPower"] = get_power_use(input_level + 1)
+
+	/// A list of lists, each inner list equals a datum
+	var/list/listed_items = list()
+	for(var/key = 1 to length(product_list))
+		var/datum/data/bluespace_tap_product/A = product_list[key]
+		listed_items[++listed_items.len] = list(
+				"key" = key,
+				"name" = A.product_name,
+				"price" = A.product_cost)
+	data["product"] = listed_items
+	return data
+
+
+/obj/machinery/power/bluespace_tap/attack_hand(mob/user)
+	add_fingerprint(user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ghost(mob/user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ai(mob/user)
+	tgui_interact(user)
+
+/**
+  * Produces the product with the desired key and increases product cost accordingly
+  */
+/obj/machinery/power/bluespace_tap/proc/produce(key)
+	if(key <= 0 || key > length(product_list))	//invalid key
+		return
+	var/datum/data/bluespace_tap_product/A = product_list[key]
+	if(!A)
+		return
+	if(A.product_cost > points)
+		return
+	points -= A.product_cost
+	A.product_cost = round(1.2 * A.product_cost, 1)
+	playsound(src, 'sound/magic/blink.ogg', 50)
+	do_sparks(2, FALSE, src)
+	new A.product_path(get_turf(src))
+
+
+
+//UI stuff below
+
+/obj/machinery/power/bluespace_tap/tgui_act(action, params)
+	if(..())
+		return 1
```

Comment:
`TRUE` here if you want to update the screen. Most tgui_acts just return here. They don't return TRUE or any value since if this happens then the UI is not interactive (or not there)